ArrayIndexOutOfBoundsException during codegen for `io:println(+~1)`

**Description:**
```
java.lang.ArrayIndexOutOfBoundsException
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
        at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
        at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:661)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generateModuleClasses(JvmPackageGen.java:418)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generate(JvmPackageGen.java:820)
        at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:92)
        at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:69)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:301)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:426)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:345)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:112)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:170)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:111)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:80)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:176)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:53)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 0
        at org.objectweb.asm.Frame.getConcreteOutputType(Frame.java:1139)
        at org.objectweb.asm.Frame.merge(Frame.java:1184)
        at org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1607)
        at org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1543)
        at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.genJMethodForBFunc(MethodGen.java:262)
        at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.generateMethod(MethodGen.java:156)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.lambda$generateModuleClasses$0(JvmPackageGen.java:467)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1746)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
        at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

Following works
```ballerina
public function main() {
   int i =  +~1;
   io:println(i);
}
```

**Steps to reproduce:**

**Affected Versions:**
SL Beta 3 RC3

**OS, DB, other environment details and versions:**

**Related Issues (optional):**
<!-- Any related issues such as sub tasks, issues reported in other repositories (e.g component repositories), similar problems, etc. -->

**Suggested Labels (optional):**
<!-- Optional comma separated list of suggested labels. Non committers can’t assign labels to issues, so this will help issue creators who are not a committer to suggest possible labels-->

**Suggested Assignees (optional):**
<!--Optional comma separated list of suggested team members who should attend the issue. Non committers can’t assign issues to assignees, so this will help issue creators who are not a committer to suggest possible assignees-->

--- a/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java
+++ b/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java
@@ -4262,11 +4262,10 @@ public class TypeChecker extends BLangNodeVisitor {
                 actualType = new BTypedescType(exprType, null);
             }
         } else {
-//            allow both addition and subtraction operators to get expected type as Decimal
-            boolean decimalNegation = OperatorKind.SUB.equals(unaryExpr.operator) && expType.tag == TypeTags.DECIMAL;
-            boolean isAdd = OperatorKind.ADD.equals(unaryExpr.operator);
-            exprType = (decimalNegation || isAdd) ? checkExpr(unaryExpr.expr, env, expType) :
-                    checkExpr(unaryExpr.expr, env);
+            //allow both addition and subtraction operators to get expected type as Decimal
+            boolean decimalAddNegate = expType.tag == TypeTags.DECIMAL &&
+                    (OperatorKind.ADD.equals(unaryExpr.operator) || OperatorKind.SUB.equals(unaryExpr.operator));
+            exprType = decimalAddNegate ? checkExpr(unaryExpr.expr, env, expType) : checkExpr(unaryExpr.expr, env);
             if (exprType != symTable.semanticError) {
                 BSymbol symbol = symResolver.resolveUnaryOperator(unaryExpr.pos, unaryExpr.operator, exprType);
                 if (symbol == symTable.notFoundSymbol) {

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/unaryoperations/UnaryExprTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/unaryoperations/UnaryExprTest.java
@@ -27,6 +27,7 @@ import org.ballerinalang.test.CompileResult;
 import org.testng.Assert;
 import org.testng.annotations.AfterClass;
 import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
 
 /**
@@ -179,14 +180,18 @@ public class UnaryExprTest {
 
     }
 
-    @Test(description = "Test complement operator")
-    public void testComplementOperator() {
-        BRunUtil.invoke(result, "testComplementOperator");
+    @Test(dataProvider = "dataToTestUnaryOperations", description = "test unary operators with types")
+    public void testUnaryOperations(String functionName) {
+        BRunUtil.invoke(result, functionName);
     }
 
-    @Test(description = "Test unary operators with int sub types")
-    public void testUnaryOperationsWithIntSubtypes() {
-        BRunUtil.invoke(result, "testUnaryOperationsWithIntSubtypes");
+    @DataProvider(name = "dataToTestUnaryOperations")
+    public Object[] dataToTestUnaryOperations() {
+        return new String[] {
+                "testComplementOperator",
+                "testUnaryOperationsWithIntSubtypes",
+                "testUnaryOperationsWithNonBasicTypes"
+        };
     }
 
     @Test(description = "Test unary statement with errors")

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/unaryoperations/UnaryExprTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/unaryoperations/UnaryExprTest.java
@@ -196,7 +196,7 @@ public class UnaryExprTest {
 
     @Test(description = "Test unary statement with errors")
     public void testUnaryStmtNegativeCases() {
-        Assert.assertEquals(resultNegative.getErrorCount(), 8);
+        Assert.assertEquals(resultNegative.getErrorCount(), 12);
         BAssertUtil.validateError(resultNegative, 0, "operator '+' not defined for 'json'", 5, 10);
         BAssertUtil.validateError(resultNegative, 1, "operator '-' not defined for 'json'", 14, 10);
         BAssertUtil.validateError(resultNegative, 2, "operator '!' not defined for 'json'", 23, 10);
@@ -209,6 +209,10 @@ public class UnaryExprTest {
                 38, 15);
         BAssertUtil.validateError(resultNegative, 7, "incompatible types: expected 'int:Unsigned8', found 'int'",
                 41, 24);
+        BAssertUtil.validateError(resultNegative, 8, "operator '~' not defined for 'decimal'", 45, 18);
+        BAssertUtil.validateError(resultNegative, 9, "operator '~' not defined for 'float'", 46, 17);
+        BAssertUtil.validateError(resultNegative, 10, "operator '~' not defined for 'decimal'", 47, 18);
+        BAssertUtil.validateError(resultNegative, 11, "operator '!' not defined for 'decimal'", 48, 18);
     }
 
     @AfterClass

--- a/tests/jballerina-unit-test/src/test/resources/test-src/expressions/unaryoperations/unary-operation-negative.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/expressions/unaryoperations/unary-operation-negative.bal
@@ -30,7 +30,7 @@ function testNotOperator() returns int {
     return i;
 }
 
-function testIncompatibleSubtypesWithUnaryOpertaors() {
+function testIncompatibleSubtypesWithUnaryOperators() {
     int:Unsigned8 x1 = -7;
     int:Signed8 x2 = +1000;
 
@@ -39,4 +39,11 @@ function testIncompatibleSubtypesWithUnaryOpertaors() {
 
     int:Unsigned8 x5 = 0;
     int:Unsigned8 x6 = ~x5;
+}
+
+function testIncompatibleUnaryOperations() {
+    decimal x1 = ~12.5d;
+    float x2 = +~12.5;
+    anydata x3 = ~12d;
+    anydata x4 = !12d;
 }

--- a/tests/jballerina-unit-test/src/test/resources/test-src/expressions/unaryoperations/unary-operation.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/expressions/unaryoperations/unary-operation.bal
@@ -185,19 +185,39 @@ function testUnaryOperationsWithNonBasicTypes() {
     int? a2 = ++i;
     int? a3 = ~i;
     int? a4 = +~i;
+    anydata a5 = +i;
+    int|float a6 = -i;
 
     float f = -5.2;
-    float? a5 = +f;
+    float? a7 = +f;
+    anydata a8 = -f;
 
     decimal d = 7.45;
-    decimal? a6 = +d;
+    decimal? a9 = +d;
+    anydata a10 = -d;
+    decimal|int a11 = +-d;
+    anydata a12 = ++d;
+
+    int:Signed16 s = 16;
+    anydata a13 = ++s;
+    int? a14 = +-s;
+    int|decimal a15 = +~s;
 
     test:assertEquals(a1, 5);
     test:assertEquals(a2, 5);
     test:assertEquals(a3, -6);
     test:assertEquals(a4, -6);
-    test:assertEquals(a5, -5.2);
-    test:assertEquals(a6, 7.45d);
+    test:assertEquals(a5, 5);
+    test:assertEquals(a6, -5);
+    test:assertEquals(a7, -5.2);
+    test:assertEquals(a8, 5.2);
+    test:assertEquals(a9, 7.45d);
+    test:assertEquals(a10, -7.45d);
+    test:assertEquals(a11, -7.45d);
+    test:assertEquals(a12, 7.45d);
+    test:assertEquals(a13, 16);
+    test:assertEquals(a14, -16);
+    test:assertEquals(a15, -17);
 }
 
 function assertEquality(anydata expected, anydata actual) {

--- a/tests/jballerina-unit-test/src/test/resources/test-src/expressions/unaryoperations/unary-operation.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/expressions/unaryoperations/unary-operation.bal
@@ -1,3 +1,5 @@
+import ballerina/test;
+
 function negativeIntTest() returns [int, int] {
     int x;
     int y;
@@ -149,6 +151,53 @@ function testUnaryOperationsWithIntSubtypes() {
     byte x17 = +7;
     assertEquality(-8, y7);
     assertEquality(7, x17);
+}
+
+function testUnaryOperationsWithNonBasicTypes() {
+    int|float x1 = +5;
+    int|float x2 = ++5;
+    int|float x3 = -5;
+    int|float x4 = --5;
+    int|decimal x5 = ~5;
+    int|float|string x6 = +~5;
+    anydata x7 = ++5;
+    anydata x8 = --5;
+    anydata x9 = +-5;
+    anydata x10 = +~5;
+    anydata x11 = ~-5;
+    anydata x12 = -~5;
+
+    test:assertEquals(x1, 5);
+    test:assertEquals(x2, 5);
+    test:assertEquals(x3, -5);
+    test:assertEquals(x4, 5);
+    test:assertEquals(x5, -6);
+    test:assertEquals(x6, -6);
+    test:assertEquals(x7, 5);
+    test:assertEquals(x8, 5);
+    test:assertEquals(x9, -5);
+    test:assertEquals(x10, -6);
+    test:assertEquals(x11, 4);
+    test:assertEquals(x12, 6);
+
+    int i = 5;
+    int? a1 = +i;
+    int? a2 = ++i;
+    int? a3 = ~i;
+    int? a4 = +~i;
+
+    float f = -5.2;
+    float? a5 = +f;
+
+    decimal d = 7.45;
+    decimal? a6 = +d;
+
+    test:assertEquals(a1, 5);
+    test:assertEquals(a2, 5);
+    test:assertEquals(a3, -6);
+    test:assertEquals(a4, -6);
+    test:assertEquals(a5, -5.2);
+    test:assertEquals(a6, 7.45d);
 }
 
 function assertEquality(anydata expected, anydata actual) {

